Framework/DataHandling/src/ISISJournal.cpp

```diff
@@ -0,0 +1,322 @@
+ * "name": "INTER00013460")
+ */
+ISISJournal::RunData createRunDataForElement(Element *element) {
+  return ISISJournal::RunData{{"name", element->getAttribute("name")}};
+}
+
+/** Extract a list of named values for the given element.
+ * @param element : the element to extract values for
+ * @param valueNames : a list of values to extract (which may be empty if
+ * nothing is requested)
+ * @param result : a map of names to values to which we add key-value pairs for
+ * each requested value
+ */
+void addValuesForElement(Element *element,
+                         std::vector<std::string> const &valueNames,
+                         ISISJournal::RunData &result) {
+  for (auto &name : valueNames)
+    result[name] = getTextValue(element->getChildElement(name));
+}
+
+/** Extract a list of named values for all child nodes in the given parent
+ * element.
+ * @param parentElement : the element containing all child nodes we want to
+ * check
+ * @param valueNames : the names of the values to extract from within the child
+ * nodes
+ * @param filters : a map of names and values to filter the results by
+ * @returns : a map of name->value pairs containing mandatory run data as well
+ * as the values that were requested
+ */
+std::vector<ISISJournal::RunData>
+getValuesForAllElements(Element *parentElement,
+                        std::vector<std::string> const &valueNames,
+                        ISISJournal::RunData const &filters) {
+  auto results = std::vector<ISISJournal::RunData>{};
+
+  auto nodeIter = TreeWalker(parentElement, NodeFilter::SHOW_ELEMENT);
+  for (auto node = nodeIter.nextNode(); node; node = nodeIter.nextSibling()) {
+    auto element = dynamic_cast<Element *>(node);
+    throwIfNotValid(element, NXENTRY_TAG);
+
+    if (matchesAllFilters(element, filters)) {
+      auto result = createRunDataForElement(element);
+      addValuesForElement(element, valueNames, result);
+      results.emplace_back(std::move(result));
+    }
+  }
+
+  return results;
+}
+
+/** Extract an attribute value for all direct children of a given element.
+ * @param parentElement : the element containing the child nodes to check
+ * @param childElementName : the name that the child elements should have
+ * @returns : the attribute values for all of the child elements
+ * @throws : if any of the child elements does not have the expected name
+ */
+std::vector<std::string>
+getAttributeForAllChildElements(Element *parentElement,
+                                const char *childElementName,
+                                const char *attributeName) {
+  auto results = std::vector<std::string>{};
+
+  auto nodeIter = TreeWalker(parentElement, NodeFilter::SHOW_ELEMENT);
+  for (auto node = nodeIter.nextNode(); node; node = nodeIter.nextSibling()) {
+    auto element = dynamic_cast<Element *>(node);
+    throwIfNotValid(element, childElementName);
+    results.emplace_back(element->getAttribute(attributeName));
+  }
```

Comment:
Good work I really like how readable this is and how everything is encapsulated above